src/dispatch/par_seq.rs

```diff
@@ -0,0 +1,406 @@
+/// #                 type SystemData = ();
+/// #                 fn run(&mut self, _: Self::SystemData) {}
+/// #             }
+/// #         )*
+/// #     };
+/// # }
+/// #
+/// # struct SysA;
+/// # struct SysB;
+/// # struct SysC;
+/// # struct SysD;
+/// # struct SysWithLifetime<'a>(&'a u8);
+/// # struct SysLocal(*const u8);
+/// #
+/// # impl_sys!(SysA SysB SysC SysD SysLocal);
+/// #
+/// # impl<'a, 'b> System<'a> for SysWithLifetime<'b> {
+/// #     type SystemData = ();
+/// #
+/// #     fn run(&mut self, _: Self::SystemData) {}
+/// # }
+///
+/// # fn main() {
+/// # #![cfg_attr(rustfmt, rustfmt_skip)]
+/// #
+/// # let pool = ThreadPool::new(Default::default()).unwrap();
+/// #
+/// # let mut res = Resources::new();
+/// let x = 5u8;
+///
+/// let mut dispatcher = ParSeq::new(
+///     seq![
+///         par![
+///             SysA,
+///             SysWithLifetime(&x),
+///             seq![
+///                 SysC,
+///                 SysD,
+///             ],
+///         ],
+///         SysB,
+///         SysLocal(&x as *const u8),
+///     ],
+///     &pool,
+/// );
+///
+/// dispatcher.dispatch(&mut res);
+/// # }
+/// ```
+pub struct ParSeq<P, T> {
+    run: T,
+    pool: P,
+}
+
+impl<P, T> ParSeq<P, T>
+where
+    P: Borrow<ThreadPool>,
+    T: for<'a> RunWithPool<'a>,
+{
+    /// Creates a new `ParSeq` dispatcher.
+    /// `run` is usually created by using the `par!` / `seq!`
+    /// macros.
+    pub fn new(run: T, pool: P) -> Self {
+        ParSeq { run, pool }
+    }
+
+    /// Dispatches the systems using `res`.
+    /// This involves zero virtual cost.
+    pub fn dispatch(&mut self, res: &mut Resources) {
+        self.run.run(res, self.pool.borrow());
+    }
+}
+
+pub trait RunWithPool<'a> {
+    fn run(&mut self, res: &'a Resources, pool: &ThreadPool);
+
+    fn reads(&self) -> Vec<ResourceId>;
+    fn writes(&self) -> Vec<ResourceId>;
+}
+
+impl<'a, T> RunWithPool<'a> for T
+where
+    T: System<'a>,
+{
+    fn run(&mut self, res: &'a Resources, _: &ThreadPool) {
+        RunNow::run_now(self, res);
+    }
+
+    fn reads(&self) -> Vec<ResourceId> {
+        use system::SystemData;
+
+        T::SystemData::reads(0)
+    }
+
+    fn writes(&self) -> Vec<ResourceId> {
+        use system::SystemData;
+
+        T::SystemData::writes(0)
+    }
+}
+
+impl<'a, H, T> RunWithPool<'a> for Par<H, T>
+where
+    H: RunWithPool<'a> + Send,
+    T: RunWithPool<'a> + Send,
+{
+    fn run(&mut self, res: &'a Resources, pool: &ThreadPool) {
+        let head = &mut self.head;
+        let tail = &mut self.tail;
+
+        let head = move || head.run(res, pool);
+        let tail = move || tail.run(res, pool);
+
+        if pool.current_thread_index().is_none() {
+            pool.join(head, tail);
+        } else {
+            join(head, tail);
+        }
+    }
+
+    fn reads(&self) -> Vec<ResourceId> {
+        let mut reads = self.head.reads();
+        reads.extend(self.tail.reads());
```

Comment:
this seems rather inefficient from the allocation point of view, assuming `tail.read()` also allocates a vector that gets thrown out